### PR TITLE
Updated Mesos formula for 0.24 release

### PR DESCRIPTION
**Pull request by haosdent**

### CODE REVIEW

**Comment by haosdent** [comment]
@massenz @vinodkone @adam-mesos

**Comment by haosdent** [comment]
Local test through https://github.com/haosdent/homebrew-mesos .

**Comment by haosdent** [comment]
Hi, @DomT4 Does jenkins have problem before?

**Comment by DomT4** [comment]
There's an installation failure:

```
Building mesos-0.24.0.jar ...
mv ../include/mesos/master/*.pb.cc master
[INFO] Scanning for projects...
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building mesos 0.24.0
[INFO] ------------------------------------------------------------------------
Downloading: https://repo.maven.apache.org/maven2/org/apache/maven/plugins/maven-remote-resources-plugin/1.3/maven-remote-resources-plugin-1.3.pom

[WARNING] Failed to create parent directories for tracking file /Users/brew/.m2/repository/org/apache/maven/plugins/maven-remote-resources-plugin/1.3/maven-remote-resources-plugin-1.3.pom.lastUpdated
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 0.346 s
[INFO] Finished at: 2015-09-20T20:06:15-07:00
[INFO] Final Memory: 6M/153M
[INFO] ------------------------------------------------------------------------
[ERROR] Plugin org.apache.maven.plugins:maven-remote-resources-plugin:1.3 or one of its dependencies could not be resolved: Failed to read artifact descriptor for org.apache.maven.plugins:maven-remote-resources-plugin:jar:1.3: Could not transfer artifact org.apache.maven.plugins:maven-remote-resources-plugin:pom:1.3 from/to central (https://repo.maven.apache.org/maven2): /Users/brew/.m2/repository/org/apache/maven/plugins/maven-remote-resources-plugin/1.3/maven-remote-resources-plugin-1.3.pom.part.lock (No such file or directory) -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginResolutionException
make[1]: *** [java/target/mesos-0.24.0.jar] Error 1
make: *** [all-recursive] Error 1
```

I haven't seen it before, but it's perhaps sandbox related.

**Comment by haosdent** [comment]
The interesting thing is I could install in my local before.  https://github.com/haosdent/homebrew-mesos I guess maven in homebrew maybe have some problems.

**Comment by haosdent** [comment]
As you see. It failed on download https://repo.maven.apache.org/maven2/org/apache/maven/plugins/maven-remote-resources-plugin/1.3/maven-remote-resources-plugin-1.3.pom while I could open it in browser.

**Comment by haosdent** [comment]
Let me use the old one (0.23) to verify if jenkins could pass not not.

**Comment by haosdent** [comment]
https://github.com/Homebrew/homebrew/pull/44199 also failed which success in https://github.com/Homebrew/homebrew/pull/42177

**Comment by xu-cheng** [comment]
Install failed probably because it tries to write files where sandbox doesn't permit. You can test this locally by `brew install mesos --sandbox --verbose`.

```
==> Sandbox log
Sep 21 01:14:41 sandboxd[1123]: java(31840) deny file-write-data /Users/travis/.oracle_jre_usage/42d8591bd1e10f3f.timestamp
Sep 21 01:14:41 sandboxd[1123]: java(31840) deny file-write-create /Users/travis/.m2
Sep 21 01:14:41 sandboxd[1123]: java(31840) deny file-write-create /Users/travis/.m2
```

**Comment by haosdent** [comment]
Hi, @xu-cheng Thank you for your reply. But "/Users/travis/.m2" this folder is always used by maven. Maven use it to store jars.

**Comment by xu-cheng** [comment]
We set `HOME` env to the `buildpath/".brew_home"` during the installation. So it's kinda strange that it tries to write files to old Home.

**Comment by DomT4** [comment]
Perhaps Java works out the `HOME` variable with another method :confused:. We have seen quite a lot of this behaviour with Java projects under the sandbox.

**Comment by haosdent** [comment]
@DomT4 Do you have any ideas to fix this?

**Comment by DomT4** [comment]
Just trying to find some Java documentation at the moment.

**Comment by DomT4** [comment]
Heh. Yeah, unless they're using shell to find out the `HOME`, Java has several internal methods that don't give a hoot about the env:

```ruby
(buildpath/"test.java").write <<-EOS.undent
  public class test
  {
    public static void main(String[] args)
    {
      System.out.println("User Home Path: "+System.getProperty("user.home"));
    }
  }
EOS
system "javac", "test.java"
system "java", "test"
```

```
==> Using the sandbox
/usr/bin/sandbox-exec -f /tmp/homebrew20150921-6710-17cwbvt.sb nice /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby -W0 -I /usr/local/Library/Homebrew -- /usr/local/Library/Homebrew/build.rb /usr/local/Library/Formula/a2ps.rb --verbose
==> javac test.java
==> java test
User Home Path: /Users/Dominyk
```

**Comment by mikemcquaid** [comment]
Apple's API calls do the same. Would be good to know if there's a better override we can set.

**Comment by DomT4** [comment]
Apparently for Java we can set this as an `ENV`:

```
_JAVA_OPTIONS=-Duser.home=$DIR
```

**Comment by haosdent** [comment]
Cool.
http://stackoverflow.com/questions/1501235/change-user-home-system-property
http://stackoverflow.com/questions/16649420/any-way-to-change-the-m2-folder-location-or-m2-conf-settings-xml-location-perm

On Mon, Sep 21, 2015 at 9:42 PM, Dominyk Tiller <notifications@github.com>
wrote:

> Apparently for Java we can set this as an ENV:
>
> _JAVA_OPTIONS=-Duser.home=$DIR
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Homebrew/homebrew/pull/44137#issuecomment-141981776>.
>



-- 
Best Regards,
Haosdent Huang

**Comment by DomT4** [comment]
```
checking for sasl_done in -lsasl2... yes
checking SASL CRAM-MD5 support... configure: error: no
-------------------------------------------------------------------
We need CRAM-MD5 support for SASL authentication.
-------------------------------------------------------------------
```

Not sure what's going on there. This is without any changes to this PR; just pulled locally & `brew install -v -fs`. Wasn't a problem on the CI.

**Comment by haosdent** [comment]
```
We need CRAM-MD5 support for SASL authentication.
```
I think it could be ignore.

**Comment by haosdent** [comment]
Does it failed again?

**Comment by DomT4** [comment]
Yup. Hard fails every time. Looking into it :sweat:

**Comment by DomT4** [comment]
Oh, one of my subversion arguments causes the build to fail. A vanilla `subversion` install is fine. Now just to work out which one.

**Comment by haosdent** [comment]
Thank you very much.

**Comment by DomT4** [comment]
The error seems to be because I didn't rebuild `subversion` after Apple updated system headers to fix security issues recently. Ho hum.

New build error:

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 02:48 min
[INFO] Finished at: 2015-09-21T15:24:14+01:00
[INFO] Final Memory: 29M/242M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:2.8.1:jar (build-and-attach-javadocs) on project mesos: MavenReportException: Error while creating archive:
[ERROR] Exit code: 1 - Picked up _JAVA_OPTIONS: -Duser.home=/private/tmp/mesos20150921-21602-19h9jrr/mesos-0.24.0/.brew_home
[ERROR] javadoc: error - No packages or classes specified.
[ERROR]
[ERROR] Command line was: /Library/Java/JavaVirtualMachines/jdk1.8.0_60.jdk/Contents/Home/bin/javadoc @options
[ERROR]
[ERROR] Refer to the generated Javadoc files in '/private/tmp/mesos20150921-21602-19h9jrr/mesos-0.24.0/src/java/target/apidocs' dir.
```

Looking into that now.

**Comment by haosdent** [comment]
Does this link help http://stackoverflow.com/questions/11683715/suppressing-the-picked-up-java-options-message

**Comment by haosdent** [comment]
alias java="`which java` -Duser.home=$DIR"

**Comment by DomT4** [comment]
Alas not. Seems to be a sandbox thing - Outside the sandbox:

```
==> Summary
🍺  /usr/local/Cellar/mesos/0.24.0: 1184 files, 78M, built in 17.4 minutes
```

Will try a few different `HOME` directories and see if any take.

**Comment by haosdent** [comment]
Thanks a lot!

**Comment by DomT4** [comment]
Thinking bug `¯\_(ツ)_/¯`. Reported [here](https://issues.apache.org/jira/browse/MESOS-3482). It seems to be working as expected, up to that point. There's no leakage and it happily writes and compiles everything to that point inside the sandbox.

**Comment by haosdent** [comment]
Thank you for your report.

**Comment by haosdent** [comment]
Do I need add _JAVA_OPTIONS to my patch and update pull request?

**Comment by DomT4** [comment]
```ruby
ENV["_JAVA_OPTIONS"] = "-Duser.home=#{ENV["HOME"]}"
```
At the top of `def install`.

**Comment by haosdent** [comment]
Thanks, let me update.

**Comment by haosdent** [comment]
@DomT4 Seems homebrew jenkins ci not work?

**Comment by DomT4** [comment]
It worked but apparently passed the wrong `HOME` env. @mikemcquaid is there a reason the bots `$HOME` wouldn't be correctly manipulated by our build env? :confused: 

```
Building mesos-0.24.0.jar ...
Picked up _JAVA_OPTIONS: -Duser.home=/Users/brew/Jenkins/prs/version/yosemite/home
[ERROR] Could not create local repository at /Users/brew/Jenkins/prs/version/yosemite/home/.m2/repository -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/LocalRepositoryNotAccessibleException
make[1]: *** [java/target/mesos-0.24.0.jar] Error 1
make: *** [all-recursive] Error 1
```

**Comment by mikemcquaid** [comment]
@DomT4 nothing special going on with the bots there.

**Comment by DomT4** [comment]
Weird. Feel free to set it manually @haosdent; `"-Duser.home=#{buildpath}/.brew_home"`. If that works across several different Java formulae might see if we can tuck it away in the core code as we do with the `$HOME` manipulation.

**Comment by haosdent** [comment]
Thank you very much, let me try again.

**Comment by haosdent** [comment]
@DomT4  Seems buildpath not exist
```
undefined local variable or method `buildpath' for #<Class:0x007fa8f40695b0>
```

**Comment by DomT4** [comment]
It needs to go just inside `def install` rather than just above `def install`.

**Comment by haosdent** [comment]
Oh, got it.

**Comment by haosdent** [comment]
Hi, @DomT4 Jenkins still failed. http://bot.brew.sh/job/Homebrew%20Pull%20Requests/33630/version=yosemite/console Is it because my code?

**Comment by DomT4** [comment]
Nope. The CI crashed. Have stuck it back in.

**Comment by haosdent** [comment]
@DomT4 Seems failed because of "_JAVA_OPTIONS"

**Comment by haosdent** [comment]
Seems the real fail reason is 
```
[ERROR] javadoc: error - No packages or classes specified.
```

Let me find the root cause of this first.

**Comment by haosdent** [comment]
This problem is a bug of maven-javadoc-plugin. https://github.com/apache/maven-plugins/blob/trunk/maven-javadoc-plugin/src/main/java/org/apache/maven/plugin/javadoc/JavadocUtil.java#L607-L611 maven-javadoc-plugin would use error output to parse javadoc version. And javadoc also outpute the version information to stdout. Because we set _JAVA_OPTIONS before, here broke javadoc version parsing.
```
        if ( StringUtils.isNotEmpty( err.getOutput() ) )
        {
            return parseJavadocVersion( err.getOutput() );
        }
```

**Comment by haosdent** [comment]
@DomT4 @xu-cheng Could we merge it now?

**Comment by mikemcquaid** [comment]
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

For future reference the preferred commit message format for simple version updates is `mesos 0.24.0`.

**Comment by haosdent** [comment]
Thank you very much @DomT4 @mikemcquaid .
